[오구] 로또 미션 제출합니다.

LTTTTTE opened the pull request:
오구입니다.
리뷰 잘부탁드립니다.

LTTTTTE committed:
docs: 기능 목록 작성

LTTTTTE committed:
feat: 구매한 로또갯수 계산하는 기능 추가

LTTTTTE committed:
feat: 랜덤수 생성하는 기능 추가

LTTTTTE committed:
feat: 수량에 맞는 로또를 생성하는 기능 추가

LTTTTTE committed:
feat: 구입 금액이 자연수인지 확인하는 기능

LTTTTTE committed:
test: Money domain initTestCode 추가

LTTTTTE committed:
feat: InputView 클래스 추가

LTTTTTE committed:
feat: 지난주 당첨번호 parse 하는 기능 추가

LTTTTTE committed:
feat: 당첨 번호의 예외 처리 기능 추가

LTTTTTE committed:
feat: 로또 번호와 보너스 번호의 중복 체크 기능 추가

LTTTTTE committed:
test: 로또 몇 등인지 확인하는 테스트 추가

LTTTTTE committed:
feat: 각 로또가 당첨번호/보너스번호 중 몇 개 일치하는지 확인하는 기능 추가

LTTTTTE committed:
feat: 당첨 통계 출력하는 기능 추가

LTTTTTE committed:
feat: 구매한 로또갯수 출력하는 기능 추가

LTTTTTE committed:
feat: 구입한 로또 출력하는 기능 추가

LTTTTTE committed:
feat: 사용자로부터 입력받는 기능 추가

LTTTTTE committed:
refactor: 구입한 로또가 정렬될수 있도록 수정

LTTTTTE committed:
refactor: 범용성을 가지도록 메소드 명 수정

LTTTTTE committed:
feat: 등수에 없는 Rank 처리하는 기능 추가

LTTTTTE committed:
feat: 메인 컨트롤러 작성

LTTTTTE committed:
test: Lotto 객체 생성 테스트 코드 추가

LTTTTTE committed:
test: NumberParser 테스트코드 작성

LTTTTTE committed:
test: BonusNumber 테스트 작성

LTTTTTE committed:
test: WinningNumber 테스트 코드 작성

LTTTTTE committed:
refactor: 패키지 분리

LTTTTTE committed:
refactor: 원시값 포장

LTTTTTE committed:
feat: 커스텀 예외처리 추가

LTTTTTE committed:
refactor: 모든 문자열 포장

LTTTTTE committed:
refactor: 메인컨트롤러 기능 분리

phs1116 commented on src/main/java/lotto/controller/LottoManager.java:
```diff
@@ -0,0 +1,28 @@
+	private static final LottoNumberGenerator lottoNumberGenerator = new LottoNumberGenerator();
+
+	public static int calculateLottoAmount(Money money) {
+		return money.getMoney() / LOTTO_PRICE;
```
getter를 통한 연산이 아닌 Money에게 메시지를 전달하는 형태로 구현해보세요!
money가 어떤 값을 가지고 있는지 객체의 구현에 대해서 최대한 캡슐화 하는걸 연습해보면 좋을 것 같아요. :)

LTTTTTE replied:
돈이 어떤값을 가지고 어떤 일을 할수있을지 생각해서 메소드 추가한후 캡슐화 해보았습니다.

phs1116 commented on src/main/java/lotto/controller/LottoManager.java:
```diff
@@ -0,0 +1,28 @@
+import lotto.domain.Money;
+import lotto.utils.LottoNumberGenerator;
+
+public class LottoManager {
```
LottoManager에 대한 테스트가 없는 것 같아요.
테스트하기 어려운 요소인 LottoNumberGenerator를 어떻게 느슨하게 의존할지 고민해보시고 테스트를 작성해주세요!

phs1116 replied:
NumberGenerator로 인터페이스를 분리하셨네요. 개방-폐쇄 원칙과 의존성 역전 원칙에 대해 공부해보세요!

LTTTTTE replied:
의존성 그래프의 최상단 Main에서 NumberGenerator를 주입할수 있도록 코드를 수정해보았습니다.
테스트가 가능한 NumberGenerator를 주입하여 테스트코드를 작성해보았습니다.

phs1116 commented on src/main/java/lotto/controller/LottoManager.java:
```diff
@@ -0,0 +1,28 @@
+import lotto.utils.LottoNumberGenerator;
+
+public class LottoManager {
+	private static final int LOTTO_PRICE = 1000;
```
LOTTO_PRICE도 Money로 볼수있지 않나요?

LTTTTTE replied:
돈을 의미하는 원시값들 모두 포장해보았습니다.

phs1116 commented on src/main/java/lotto/controller/RankCalculator.java:
```diff
@@ -0,0 +1,36 @@
+			.collect(Collectors.toList());
+	}
+
+	public static Rank calculateSingle(Lotto lotto, List<Integer> winningNumbers, int bonusNumber) {
```
List<Integer> winningNumbers, int bonusNumber도 WinningNumber인데 실수하신것같아요!

phs1116 commented on src/main/java/lotto/controller/RankCalculator.java:
```diff
@@ -0,0 +1,36 @@
+			.stream()
+			.filter(winningNumbers::contains)
+			.count();
+	}
```
getter를 통해 값을 꺼내 계산하는 것이 아닌, WinningNumber와 Lotto의 상호작용을 통해 구현해보도록 해보면 좋을 것 같아요. :)

getter를 필요한 경우 아니면 최대한 지양하도록 노력해보세요.

LTTTTTE replied:
객체간 상호작용을 통해 getter 를 최대한 없애보았습니다.

phs1116 commented on src/main/java/lotto/domain/BonusNumber.java:
```diff
@@ -0,0 +1,26 @@
+public class BonusNumber {
+	private static final int MAX_LOTTO_NUMBER_RANGE = 45;
+	private static final int MIN_LOTTO_NUMBER_RANGE = 1;
+	private static final String LOTTO_NUMBER_RANGE_MESSAGE = "1~45 범위의 숫자만 로또 번호가 될 수 있습니다.";
```
해당 제약 조건은 BonusNumber만 가지는 것이 아닌 로또 번호 그 자체에게 필요한 제약으로 보여요.
로또 번호 그 자체를 값객체로 분리해보면 어떨까요?

LTTTTTE replied:
번호 자체 값을 가지는 LottoNumber 객체를 새로 만들어 적용해보았습니다.

phs1116 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,43 @@
+	private static final int MIN_LOTTO_NUMBER_RANGE = 1;
+	private static final int LOTTO_NUMBER_SIZE = 6;
+	private static final String LOTTO_NUMBER_RANGE_MESSAGE = "1~45 범위의 숫자만 로또 번호가 될 수 있습니다.";
+	private static final String LOTTO_NUMBER_DUPLICATED_MESSAGE = "잘못된 로또 번호입니다. 중복 안됨, 갯수는 6개";
```
BonusNumber와 같은 로직이 중복이네요. 분리해보면 좋을 것 같습니다. :)

LTTTTTE replied:
LottoNumber 객체를 생성함으로써 중복을 없애보았습니다.

phs1116 commented on src/main/java/lotto/domain/WinningNumber.java:
```diff
@@ -0,0 +1,31 @@
+	}
+
+	private boolean bonusNumberDuplicatedWithWinningNumber(BonusNumber bonusNumber) {
+		return winningNumber.getNumbers().contains(bonusNumber.getBonusNumber());
```
RankCalculator에 드린 피드백을 참고하면 이런 부분도 개선 될 수 있을 것 같아요!

LTTTTTE replied:
객체에 메세지를 보내는 방식으로 개선해보았습니다.

phs1116 commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -0,0 +1,65 @@
+			.filter(Objects::nonNull)
+			.mapToInt(Rank::getReward)
+			.sum();
+		System.out.printf(EARNING_RATE_MESSAGE, (double)sum / money.getMoney() * 100);
```
100 하드코딩이 아닌 상수로. :)

LTTTTTE replied:
하드코딩 상수화 하였습니다.

phs1116 commented on src/test/java/lotto/domain/BonusNumberTest.java:
```diff
@@ -0,0 +1,18 @@
+		BonusNumber bonusNumber = new BonusNumber(1);
+		assertThat(bonusNumber).isNotNull();
+		assertThatThrownBy(() -> new BonusNumber(46))
+			.isInstanceOf(LottoNumberRangeException.class)
```
예외 케이스, 정상 케이스에 대해서는 테스트를 분리하는게 좋아요.  :)
테스트 명만 보고 어떤 케이스가 실패했는지 알수 있도록!

LTTTTTE replied:
예외 케이스와 정상 케이스 별로 분리 해보았습니다.

phs1116 reviewed (changes requested):
안녕하세요, 오구!
리뷰어 화투에요.
전체적으로 구현 매우 잘 해주셨어요. :)
피드백 추가했는데 참고 부탁드리고
궁금하신 점 있으시면 연락주세요!

LTTTTTE committed:
refactor: 피드백 반영

- 객체에 get 하지 않고 메세지를 전달하도록 수정
- 돈을 의미하는 원시값들 포장
- LottoNumber 로 포장함으로써 중복코드 제거
- 하드코딩 상수화
- LottoManager 테스트 코드 추가

phs1116 commented on src/main/java/lotto/controller/RankCalculator.java:
```diff
@@ -0,0 +1,22 @@
+import lotto.domain.WinningNumber;
+
+public class RankCalculator {
+	public static List<Rank> calculateMultiple(List<Lotto> lottos, WinningNumber winningNumber) {
+		return lottos.stream()
+			.map(x -> calculateSingle(x, winningNumber))
+			.collect(Collectors.toList());
+	}
+
+	public static Rank calculateSingle(Lotto lotto, WinningNumber winningNumber) {
+		int hitNumber = winningNumber.countHitNumber(lotto);
+		boolean bonusNumberExist = winningNumber.hasBonusNumber(lotto);
+		return Rank.getRank(hitNumber, bonusNumberExist);
+	}
```
이제 이렇게 분리하고 보니 해당 책임을 이제 winningNumber나 Lotto가 가져가도 되지 않을까요? :)
winningNumber.match(lotto) -> Rank :)

phs1116 commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -0,0 +1,43 @@
+
+import lotto.exceptions.NotAllowedMoneyAmountException;
+
+public class Money {
```
Money를 값객체로 활용하신다면 동등성을 보장하기 위해 equals, hashcode 구현해주시는게 좋아요. :)

phs1116 commented on src/main/java/lotto/controller/LottoManager.java:
```diff
@@ -0,0 +1,27 @@
+package lotto.controller;
```
해당 코드는 domain에 더 가깝지 않을까요?

phs1116 reviewed (approved):
안녕하세요, 오구!
리뷰어 화투입니다.
매우 깔끔하게 구현 잘해주셨어요. :) 
이번 단계는 머지할게요.
몇가지 간단한 피드백 추가했는데 다음 단계 진행하시면서 참고 부탁드릴게요!
